Updates coalesce.go to use log/slog and update tests to pass

jamoflaw opened the pull request:
This adds `log/slog` to the `coalesce.go` file and also defines a slog handler to capture log records as they are written by slog for validation in tests.

- [ ] this PR contains user facing changes (the `docs needed` label should be applied if so)
- [x] this PR contains unit tests
- [x] this PR has been tested for backwards compatibility

jamoflaw committed:
Updatse coalesce.go to use log/slog and update tests to pass

Signed-off-by: James Payne <jamoflaw@gmail.com>

jamoflaw commented on pkg/chart/v2/util/coalesce_test.go:
```diff
@@ -702,22 +702,26 @@ func TestCoalesceValuesWarnings(t *testing.T) {
-	assert.Contains(t, warnings, "warning: skipped value for level1.level2.level3.boat: Not a table.")
-	assert.Contains(t, warnings, "warning: destination for level1.level2.level3.spear.tip is a table. Ignoring non-table value (true)")
-	assert.Contains(t, warnings, "warning: cannot overwrite table with non table for level1.level2.level3.spear.sail (map[cotton:true])")
+	assert.Contains(t, logOutput, "warning: skipped value for level1.level2.level3.boat: Not a table.")
```
at this point none of the log message output has been altered, though there is a defined method `AsMessageLevelMap` which should later allow comparison against the message _and_ level for tests and in theory the slog attrs depending how deep the tests should go

jamoflaw commented on pkg/chart/v2/util/utilities_test.go:
```diff
@@ -0,0 +1,131 @@
+limitations under the License.
+*/
+
+package util
```
This seemed odd to me - the other tests in this package are defined under `util` and not `util_test` - was there any reason for that or is it just historical?

benoittgt commented:
I have a question about our logging approach: should we standardize on key-value rather than string interpolation for these warning logs? 

While key-value format offers better structure, these are warning logs (not debug logs) so volume should be lower. 

Perhaps we should establish formal logging guidelines for this repository to ensure consistency? 

cc @robertsirc @scottrigby

TerryHowe commented:
> I have a question about our logging approach: should we standardize on key-value rather than string interpolation for these warning logs?

Using slog without key values is kind of pointless so I'd think these should be changed

jamoflaw committed:
Remove erroneous use of Sprintf

Signed-off-by: James Payne <jamoflaw@gmail.com>

jamoflaw commented:
> > I have a question about our logging approach: should we standardize on key-value rather than string interpolation for these warning logs?
> 
> Using slog without key values is kind of pointless so I'd think these should be changed

This definitely makes sense

A standardised approach for the k:v pairs I think would be a useful thing - I'm going to have a play around with the information which is currently available to add some context to these messages, just so there is something to discuss

jamoflaw committed:
Update logging in coalesce.go to use structured key:pairs.

Update logger test framework to handle more test cases using a builder pattern for easier reuse

Signed-off-by: James Payne <jamoflaw@gmail.com>

jamoflaw commented on pkg/chart/v2/util/coalesce.go:
```diff
@@ -129,20 +129,20 @@ func coalesceDeps(printf printFn, chrt *chart.Chart, dest map[string]interface{}
 		dg = make(map[string]interface{})
 	} else if dg, ok = destglob.(map[string]interface{}); !ok {
-		printf("warning: skipping globals because destination %s is not a table.", GlobalKey)
+		log.Warn(fmt.Sprintf("skipping globals because destination %s is not a table.", GlobalKey), slog.String("key", GlobalKey), slog.String("error", "not a table"))
```
I've added keys for the `key` in values, and the error msg (given these are warnings is `error` the right key here?)

Where possible I've added the chart name for some context - though I think it needs to be seen how much value that would add

jamoflaw commented on pkg/chart/v2/util/coalesce_test.go:
```diff
@@ -702,22 +702,51 @@ func TestCoalesceValuesWarnings(t *testing.T) {
-	assert.Contains(t, warnings, "warning: cannot overwrite table with non table for level1.level2.level3.spear.sail (map[cotton:true])")
 
+	// All warnings should have context as to where the warning is being emitted
+	capturedLogOutput.Filter(RecordLevelMatches(slog.LevelWarn)).
```
I reworked the test framework to be more useful for validating logs are to some sort of standard. You can filter by level, message (or whatever else needed) and then assert particular keys are present (so here just checking the 3 messages which were there previous have keys set on them)

jamoflaw committed:
Remove function which is no longer used

Signed-off-by: James Payne <jamoflaw@gmail.com>

jamoflaw committed:
Update log messages to better leverage the slog k:v pairs
Update log messages to use consistent language throughout

Signed-off-by: James Payne <jamoflaw@gmail.com>

jamoflaw commented:
This is now updated to use:

- slog k:v pairs throughout rather than sprinkling `fmt.Sprintf` in places
- language has been made consistent in all messages

Is this along correct lines here? Just thinking about overall direction really - specifically around the testing methods

I think there are still a few opportunities to add more logs (I'm thinking around the nil / delete removals)?

TerryHowe commented on pkg/chart/v2/util/coalesce.go:
```diff
@@ -133,14 +133,16 @@ func coalesceGlobals(log *slog.Logger, dest, src map[string]interface{}, prefix
 	} else if dg, ok = destglob.(map[string]interface{}); !ok {
-		log.Warn(fmt.Sprintf("skipping globals because destination %s is not a table.", GlobalKey), slog.String("key", GlobalKey), slog.String("error", "not a table"))
+		// Destination is not a table
+		log.Warn("skipping coalescing global values", slog.String("key", GlobalKey), slog.String("error", "key is not a table"))
```
Very confusing that the slog.logger is called log

jamoflaw replied:
yes possibly naming it `logger` might be more appropriate? Incase its not clear its passed through from the calling methods to allow it to inherit the `chart` key - if that key is dropped then it can just use the standard slog interface

jamoflaw replied:
I've updated to be `logger` in those areas now

jamoflaw replied:
```
~/Documents/Projects/helm$ helm template testing ./test/chart1/ | head -n 10
level=WARN msg="skipping coalescing global values" chart=chart1 key=global error="key is not a table"
level=WARN msg="skipping key" chart=chart2 key=chart1.chart2.outer error="cannot merge table and non-table values"
level=WARN msg="skipping coalescing global values" chart=chart1 key=global error="key is not a table"
level=WARN msg="skipping key" chart=chart2 key=chart1.chart2.outer error="cannot merge table and non-table values"
level=WARN msg="skipping coalescing global values" chart=chart1 key=global error="key is not a table"
level=WARN msg="skipping key" chart=chart2 key=chart1.chart2.outer error="cannot merge table and non-table values"
---
# Source: chart1/charts/chart2/templates/serviceaccount.yaml
apiVersion: v1
kind: ServiceAccount
```

I think its debatable if the chart key adds enough additional context?

as it also gets called from the exported functions:

```
func CoalesceTables(dst, src map[string]interface{}) map[string]interface{} {
	return coalesceTablesFullKey(nil, dst, src, "", false)
}

func MergeTables(dst, src map[string]interface{}) map[string]interface{} {
	return coalesceTablesFullKey(nil, dst, src, "", true)
}```

where currently a nil logger is passed in, to get the chart key in there would mean updating the exported function signature - if the `chart` name key is seen as a useful addition then I would think it be worth updating to pass it (the chart object?) through also

jamoflaw committed:
Update pointer name to differentiate it from the log package

Signed-off-by: James Payne <jamoflaw@gmail.com>

jamoflaw committed:
Move test framework into sharable internal package

Signed-off-by: James Payne <jamoflaw@gmail.com>

TerryHowe commented on pkg/chart/v2/util/coalesce.go:
```diff
@@ -114,10 +112,10 @@ func coalesceDeps(printf printFn, chrt *chart.Chart, dest map[string]interface{}
 			subPrefix := concatPrefix(prefix, chrt.Metadata.Name)
 			// Get globals out of dest and merge them into dvmap.
-			coalesceGlobals(printf, dvmap, dest, subPrefix, merge)
+			coalesceGlobals(slog.Default().With(slog.String("chart", chrt.Name())), dvmap, dest, subPrefix, merge)
```
Sorry my reviews aren't complete I'm traveling and on my phone atm. Anyway the approach in the other PRs was to remove the passes in logger and call slog.Warn directly

jamoflaw replied:
No problem at all!

I've removed the passed in child logger completely for now to simplify things slightly - so this just now updates the log message language to be consistent and updates the log testing framework to be generic for slog and reusable

jamoflaw committed:
Remove passed in child logger completely

Signed-off-by: James Payne <jamoflaw@gmail.com>

benoittgt commented:
I like this last version. :) Good work.

jamoflaw committed:
Fix tests after child logger was removed

Signed-off-by: James Payne <jamoflaw@gmail.com>

jamoflaw commented:
Just fixed the tests after removing the child logger (some areas don't have access to the chart name now so had to remove the checks for the chart attr in places)

I ran the tests this time locally and they pass again! (sorry @gjenkins8!)

TerryHowe commented on pkg/chart/v2/util/coalesce.go:
```diff
@@ -129,20 +127,22 @@ func coalesceDeps(printf printFn, chrt *chart.Chart, dest map[string]interface{}
 	} else if dg, ok = destglob.(map[string]interface{}); !ok {
-		printf("warning: skipping globals because destination %s is not a table.", GlobalKey)
+		// Destination is not a table
+		slog.Warn("skipping coalescing global values", slog.String("key", GlobalKey), slog.String("error", "key is not a table"))
```
I haven't tried it, but does it produce the same thing if you don't specify `slog.String` for these?

jamoflaw replied:
It does indeed, those helpers are a mechanism to compile time enforce pairs for the KV elements. I believe they are also strongly typed as there is an interesting variant too for numerics

Without that the pairs can, in theory, become unbalanced if someone deleted one half, and it would still compile but produce !BADKEY as the value for one of the pairs only visible at runtime
